url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Visualizer.java

```diff
@@ -0,0 +1,547 @@
+				return i;
+		
+		return null;
+	}
+	
+	/**
+	 * Shows an option dialog with a droplist for the user to select from a certain range of options.
+	 * @param title the title of the dialog
+	 * @param description the description, which should be longer than the title
+	 * @param options the possible options that would appear in the droplist
+	 * @return the id of the option that was selected or -1 if nothing was selected
+	 */
+	public int showOptionDialog(final String title, final String description, final Object[] options) {
+		final Object ans = JOptionPane.showInputDialog(
+                this,
+                description,
+                title,
+                JOptionPane.PLAIN_MESSAGE,
+                null,
+                options,
+                options[0].toString());
+		if (ans == null) return -1;
+		for (int i = 0; i < options.length; i++) if (ans == options[i]) return i;
+		return -1;
+	}
+	
+	/**
+	 * Shows an input dialog that the user can type information in.
+	 * @param title the title of the dialog
+	 * @param description the description, which should be longer than the title
+	 * @param default_text the default text to be shown in the dialog
+	 * @return
+	 */
+	public String showInputDialog(final String title, final String description, final String default_text) {
+		return (String) JOptionPane.showInputDialog(
+                this,
+                description,
+                title,
+                JOptionPane.PLAIN_MESSAGE,
+                null,
+                null,
+                default_text);
+	}
+	
+	/**
+	 * Gets all of the items from the given type
+	 * @param arr
+	 * @return
+	 */
+	@SuppressWarnings("unchecked")
+	public <T> T[] getAllItemsOfType(final T[] arr) {
+		
+		final Class<?> input = arr.getClass().getComponentType();
+		final ArrayList<T> al = new ArrayList<T>();
+		
+		for (final Item it : items)
+			if (it.getClass().equals(input))
+				al.add((T) it);
+		
+		return al.toArray(arr);
+	}
+	
+	/**
+	 * Gets all of the qubits that have pointers starting from them pointing to the qubit q
+	 * @param q
+	 * @param type
+	 * @return
+	 */
+	public Qubit[] getQubitsDepending(final Qubit q, final Corrector.corrtype type) {
+		final ArrayList<Qubit> qubits = new ArrayList<Qubit>();
+		final Corrector[] correctors = getAllItemsOfType(new Corrector[0]);
+		
+		for (final Corrector c : correctors)
+			if (c.t == type && c.i2.id == q.id)
+				qubits.add(c.i1);
+		
+		return qubits.toArray(new Qubit[0]);
+	}
+	
+	private String getDependance(final Qubit q, final Corrector.corrtype type, final String var) {
+		
+		
+		final Qubit[] dep = getQubitsDepending(q, type);
+		
+		if (dep.length == 0) return "0";
+		
+		final StringBuilder sb = new StringBuilder();
+		
+		sb.append(var + Tools.zBOB(dep[0].getId()));
+		
+		for (int i = 1; i < dep.length; i++)
+			sb.append("+"+var+Tools.zBOB(dep[i].getId()));
+		
+		return sb.toString();
+	}
+	
+	public McalcDescription generateMcalcDesc() throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 393-413, lines 416-425, lines 430-439, lines 444-470 into separate methods.